dora/src/cannon/codegen.rs

```diff
@@ -938,6 +990,191 @@ where
+
+        let cls = self.vm.class_defs.idx(class_def_id);
+        let cls = cls.read();
+
+        self.asm.emit_comment(Comment::Alloc(class_def_id));
+
+        let alloc_size = match cls.size {
+            InstanceSize::Fixed(size) => AllocationSize::Fixed(size as usize),
+            _ => unimplemented!(
+                "class size type {:?} for new object not supported",
+                cls.size
+            ),
+        };
+
+        let array_ref = match cls.size {
+            InstanceSize::ObjArray => true,
+            _ => false,
+        };
+
+        let gcpoint = GcPoint::from_offsets(self.references.clone());
+        self.asm.allocate(
+            REG_RESULT.into(),
+            alloc_size,
+            Position { line: 0, column: 0 }, // Correct position is currently not known.
+            array_ref,
+            gcpoint,
+        );
+
+        let bytecode_type = self.bytecode.register_type(dest);
+        let offset = self.bytecode.register_offset(dest);
+
+        // store gc object in temporary storage
+        self.asm
+            .store_mem(bytecode_type.mode(), Mem::Local(offset), REG_RESULT.into());
+
+        // store classptr in object
+        let cptr = (&**cls.vtable.as_ref().unwrap()) as *const VTable as *const u8;
+        let disp = self.asm.add_addr(cptr);
+        let pos = self.asm.pos() as i32;
+
+        self.asm.emit_comment(Comment::StoreVTable(class_def_id));
+        self.asm.load_constpool(REG_TMP1.into(), disp + pos);
+        self.asm
+            .store_mem(MachineMode::Ptr, Mem::Base(REG_RESULT, 0), REG_TMP1.into());
+
+        // clear mark/fwdptr word in header
+        assert!(Header::size() == 2 * mem::ptr_width());
+        self.asm.load_int_const(MachineMode::Ptr, REG_TMP1, 0);
+        self.asm.store_mem(
+            MachineMode::Ptr,
+            Mem::Base(REG_RESULT, mem::ptr_width()),
+            REG_TMP1.into(),
+        );
+
+        match cls.size {
+            InstanceSize::Fixed(size) => {
+                self.asm.fill_zero(REG_RESULT, size as usize);
+            }
+            _ => {}
+        }
+
+        self.references.push(offset);
+    }
+
+    fn emit_throw(&mut self, exception: Register) {
+        assert_eq!(self.bytecode.register_type(exception), BytecodeType::Ptr);
+
+        let bytecode_type = self.bytecode.register_type(exception);
+        let offset = self.bytecode.register_offset(exception);
+
+        self.asm
+            .load_mem(bytecode_type.mode(), REG_RESULT.into(), Mem::Local(offset));
+
+        self.asm.throw(
+            REG_RESULT,
+            Position { line: 0, column: 0 }, // Correct position is currently not known.
+        );
+    }
+
+    fn emit_invoke_direct_void(&mut self, fct_id: FctId, start_reg: Register, num: u32) {
+        assert_eq!(self.bytecode.register_type(start_reg), BytecodeType::Ptr);
+
+        let _fct = self.vm.fcts.idx(fct_id);
+        let _fct = _fct.read();
```

Comment:
Here you could add an assert that the function does not have type params.

Reply:
An assert for the class type params would also make sense, wouldn't it?

Reply:
Yes